sklearn/calibration.py

```diff
@@ -25,6 +27,230 @@
+                                      self.pos_label,
+                                      self.min_val_tnr,
+                                      self.min_val_tpr).fit(
+                        X[test], y[test]
+                    ).threshold_
+                )
+            self.threshold_ = sum(thresholds) / len(thresholds)
+            self.base_estimator.fit(X, y)
+        return self
+
+    def predict(self, X):
+        X = check_array(X)
+        check_is_fitted(self, ["label_encoder_", "threshold_"])
+
+        return self.label_encoder_.inverse_transform(
+            (self.base_estimator.predict_proba(X)[:, self.pos_label] >
+             self.threshold_).astype(int)
+        )
+
+
+class _CutoffClassifier(object):
+    """Optimal cutoff point selection.
+
+    It assumes that base_estimator has already been fit, and uses the input set
+    of the fit function to select an optimal cutoff point. Note that this
+    class should not be used as an estimator directly. Use the
+    OptimalCutoffClassifier with cv="prefit" instead.
+
+    Parameters
+    ----------
+    base_estimator : instance BaseEstimator
+        The classifier whose decision threshold will be adapted according to
+        the acquired optimal cutoff point
+
+    method : 'roc' or 'max_tpr' or 'max_tnr'
+        The method to use for choosing the cutoff point.
+
+    pos_label : object
+        Label considered as positive during the roc_curve construction.
+
+    min_val_tnr : float in [0, 1]
+        minimum required value for true negative rate (specificity) in case
+        method 'max_tpr' is used
+
+    min_val_tpr : float in [0, 1]
+        minimum required value for true positive rate (sensitivity) in case
+        method 'max_tnr' is used
+
+    Attributes
+    ----------
+    threshold_ : float
+        Acquired optimal decision threshold for the positive class
+    """
+    def __init__(self, base_estimator, method, pos_label, min_val_tnr,
+                 min_val_tpr):
+        self.base_estimator = base_estimator
+        self.method = method
+        self.pos_label = pos_label
+        self.min_val_tnr = min_val_tnr
+        self.min_val_tpr = min_val_tpr
+
+    def fit(self, X, y):
+        """Select a decision threshold for the fitted model's positive class
+        using one of the available methods
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data.
+
+        y : array-like, shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+            Instance of self.
+        """
+        y_score = self.base_estimator.predict_proba(X)[:, self.pos_label]
+        fpr, tpr, thresholds = roc_curve(y, y_score, self.pos_label)
+
+        if self.method == 'roc':
+            self.threshold_ = thresholds[np.argmin(
+                euclidean_distances(np.column_stack((fpr, tpr)), [[0, 1]])
```

Comment:
This is the same as min of `fpr**2 + (tpr-1)**2` which is cheaper to compute and doesn't obfuscate the simplicity of the calculation